src/expressions/operator-expr.md

```diff
@@ -360,6 +360,18 @@ Here `*T` means either `*const T` or `*mut T`.
     * `char` casts to the value of the code point, then uses a numeric cast if needed.
 * `u8` to `char` cast
     * Casts to the `char` with the corresponding code point.
+* Pointer to address cast
+    * Casting from a valid raw pointer to `usize` will produce the address that is pointed to.
+
+      The pointer's provenance is lost in this conversion.
+* Address to pointer cast
+    * Casting from `usize` to a raw pointer will produce a raw pointer to the same location as the original pointer, if the `usize` was obtained through a pointer to address cast of a valid pointer.
+
+      <div class="warning">
+      Warning:
+      The two pointers are not equivalent.
+      Dereferencing the pointer obtained from the address to pointer cast may be <a href="../behavior-considered-undefined.md">undefined behavior</a> if aliasing rules are not followed.
```

Comment:
FWIW, the goal is to make dereferencing the cast-back ptr *at least as correct* as using the original. Otherwise, int-to-ptr casts seem rather useless. As written here, I don't see how this even helps anyone since there is no way to ensure that the result of the int-to-ptr cast is usable.

Also it would really help to introduce some shorter names than "the pointer obtained from the address to pointer cast".^^ I'd probably say something like

> Casting `x: usize` to a raw pointer `ptr`, where `x` was obtained by a pointer to address cast of some raw pointer `orig_ptr`, will produce a raw pointer to the same location as `orig_ptr`.

Now later text can refer to `ptr` and `orig_ptr`. (Names are bikesheddable of course.)

Reply:
I think the warning may want to note that the aliasing rules are still in flux? Or perhaps it says that are the linked location.